src/Raven.Server/Documents/AI/Embeddings/EmbeddingsBatchingWorker.cs

```diff
@@ -0,0 +1,262 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.Embeddings;
+using Raven.Server.Background;
+using Raven.Server.Config.Categories;
+using Raven.Server.Documents.ETL.Providers.AI;
+using Sparrow.Server.Logging;
+
+namespace Raven.Server.Documents.AI.Embeddings
+{
+    public class EmbeddingsBatchingWorker(string databaseName,
+        AiConfiguration configuration,
+#pragma warning disable SKEXP0001
+        ITextEmbeddingGenerationService service,
+#pragma warning restore SKEXP0001
+        AiConnectionStringIdentifier connectionStringId,
+        SemaphoreSlim concurrencyLimiter,
+        RavenLogger logger,
+        CancellationToken shutdown)
+        : BackgroundWorkBase(databaseName, logger, shutdown)
+    {
+        private readonly string _databaseName = databaseName;
+
+        private readonly ConcurrentQueue<EmbeddingsBatchRequest> _requestQueue = new();
+        private readonly Stopwatch _batchTimer = new();
+
+        public Task<ReadOnlyMemory<float>[]> EnqueueRequestAsync(IList<string> values, CancellationToken cancellationToken)
+        {
+            var request = new EmbeddingsBatchRequest(values, callerToken: cancellationToken, workerToken: CancellationToken);
+
+            bool wasEmpty = _requestQueue.IsEmpty;
+            _requestQueue.Enqueue(request);
+
+            // Start the batch timer if this is the first request
+            if (wasEmpty)
+                _batchTimer.Restart();
+
+            return request.TaskCompletionSource.Task;
+        }
+
+        protected override async Task DoWork()
+        {
+            try
+            {
+                // Check if we have requests to process
+                if (_requestQueue.IsEmpty)
+                {
+                    // No requests, wait for a bit before checking again
+                    await WaitOrThrowOperationCanceled(TimeSpan.FromMilliseconds(10));
+                    return;
+                }
+
+                if (_batchTimer.IsRunning == false)
+                    _batchTimer.Restart();
+
+                // Nagle's algorithm:
+                // 1. If we have a full batch, process immediately
+                // 2. If the timeout has elapsed, process immediately
+                // 3. Otherwise, wait for more requests longer
+                bool shouldProcessNow = _requestQueue.Count >= configuration.MaxBatchSize ||
+                                        _batchTimer.ElapsedMilliseconds >= configuration.BatchTimeoutInMs;
```

Comment:
I have some doubts about "Otherwise, wait for more requests longer" comment. This will slow down the queries unnecessary. On a system which isn't heavy loaded it will add extra 200ms (`configuration.BatchTimeoutInMs`) to each query - do I understand it correctly? Maybe the default should be 0ms?

Reply:
In my opinion, that's the whole point of Nagle's algorithm: to slightly sacrifice the speed of single requests in favor of overall speed and reducing the request frequency to avoid violating rate limits.

The default value of 200ms is probably too high, and it might be worth considering a smaller value of around 75-100ms, but I believe that this value should be different from 0ms by default (which essentially disables batching).